### PR TITLE
174 incremental c

### LINKED ISSUE #174: incremental c
todo:
 - [ ] coordinate with #153 to register incremental c artefacts as "current"

### PR DESCRIPTION
**Pull request by bblay**
Incremental C building.
- Function to get C compiler largely the same as for Fortran, so now in `util.get_tool()`.
- Added `get_fortran_compiler()` for user convenience.
- Added `-c` if missing, with warning. This seems like a step towards C compiler awareness, like we have with Fortran.
- Moved `get_compiler_version()` to util because C compiler version format matches Fortran.

todo:
 - [ ] coordinate with #153 to register incremental c artefacts as "current"

Closes #174

### CODE REVIEW

File: source/fab/steps/compile_fortran.py [hunk 1]
```diff
@@ -218,8 +214,6 @@ def process_file(self, analysed_file: AnalysedFile):
             Before compiling a file, we calculate the combo hashes and see if the output files already exists.
 
         """
-        # todo: include compiler version in hashes
```
Comment (MatthewHambley) [inline source/fab/steps/compile_fortran.py:221]: Has this been done? `_get_obj_combo_hash` doesn't seem to do it. Is it no longer relevant?

File: source/fab/steps/compile_c.py [hunk 2]
```diff
@@ -56,7 +57,19 @@ def __init__(self, compiler: Optional[str] = None, common_flags: Optional[List[s
         """
         super().__init__(name=name)
 
-        self.compiler = compiler or os.getenv('CC', 'gcc -c')
+        self.compiler, compiler_flags = get_tool(compiler or os.getenv('CC', 'gcc -c'))  # type: ignore
```
Comment (MatthewHambley) [inline source/fab/steps/compile_c.py:60]: Should there be a `get_c_compiler` to mirror the `get_fortran_compiler`? Or should the latter be replaced by `get_tool`?

**Review by MatthewHambley — CHANGES_REQUESTED** [review]
A few questions regarding the extent of this change.

File: source/fab/steps/compile_fortran.py (hunk 1 above)
Comment (bblay) [reply source/fab/steps/compile_fortran.py:221]: Yes, done and merged in #150. It's on line 267, in `_get_obj_combo_hash`.

File: source/fab/steps/compile_c.py (hunk 2 above)
Comment (bblay) [reply source/fab/steps/compile_c.py:60]: We could definitely remove `get_fortran_compiler` and just call `get_tool(compiler or os.getenv('FC', ''))`. In fact this PR did that at one point. The reason I added `get_fortran_compiler` was as a convenience function for user configs. I think it improves ease and readability in user code. Happy to remove if preferred, of course.

File: source/fab/steps/compile_fortran.py (hunk 1 above)
Comment (bblay) [reply source/fab/steps/compile_fortran.py:221]: (now on line 145, post merge)

File: source/fab/steps/compile_c.py (hunk 2 above)
Comment (MatthewHambley) [reply source/fab/steps/compile_c.py:60]: At some point people will start to use C++ modules. That will precipitate exactly the same issues surrounding Fortran modules. We will then need per-compiler builds. In the meantime we can leave it as is.

**Review by MatthewHambley — APPROVED** [review]
I am happy that we know where the limitations are.

**Pull request merged**